### PR TITLE
Added an api for parsing delimited values

### PR DESCRIPTION
**Pull request by etorreborre**
New code...

I basically pulled out a few parsers from Javaland and a bunch of test cases from 2 websites.

It turns out that no library I have used passes all the test cases whereas my hand-rolled (but terribly ugly) implementation + a Parboiled-based one are both passing.

Next thing is to:

 - remove the useless implementations
 - benchmark Parboiled vs hand-rolled
 - add a few missing test cases from the csv-spectrum cases

### CODE REVIEW

File: mundane-dsv/src/main/scala/com/ambiata/mundane/dsv/DelimitedValuesParser.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+  def parse(line: String): String \/ List[String] =
+    \/.fromTryCatchNonFatal(
+      CSVParser.parse(line, format).getRecords.toList.headOption.map(_.iterator.toList).getOrElse(Nil)).leftMap(_.getMessage)
+}
+
+
+sealed trait DelimitedValuesParserImplementation
+
+object CommonsCsv extends DelimitedValuesParserImplementation
```
Comment (charleso) [inline mundane-dsv/src/main/scala/com/ambiata/mundane/dsv/DelimitedValuesParser.scala]: I missed half the conversation, so apologies if this was talked about then. Are we likely to ever have consumers use different implementations?

If we are can we just have `DelimitedValuesParser.createParserXYZ` functions? How much are we going to be passing the implementation type around?

File: mundane-dsv/src/test/scala/com/ambiata/mundane/dsv/DelimitedValuesParserSpec.scala [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  def malformed = prop { csvLine: CsvMalformedLine =>
+    DelimitedValuesParser.csv.parse(csvLine.value) must be_-\/
+  }
+
+  case class CsvLine(value: String, expected: List[String])
+
+  implicit def ArbitraryCsvLine: Arbitrary[CsvLine] =
+    Arbitrary(Gen.sized(n => listOfN(n, genRegularField).map(fs => CsvLine(fs.mkString(","), fs))))
+
+  case class CsvQuotedLine(value: String, expected: List[String])
```
Comment (charleso) [inline mundane-dsv/src/test/scala/com/ambiata/mundane/dsv/DelimitedValuesParserSpec.scala]: FWIW I wouldn't both fields here, just because one of them is derived and prints out if the test fails and can get in the way.

File: project/build.scala [hunk 3]
```diff
@@ -12,7 +12,7 @@ object build extends Build {
     , settings = standardSettings ++ promulgate.library("com.ambiata.mundane", "ambiata-oss")
     , aggregate = Seq(bytes, cli, control, data, error, io, parse, reflect, ssh, testing, testingExtra, time, trace)
     )
-    .dependsOn(bytes, cli, control, data, error, io, parse, reflect, ssh, testing, time, trace)
+    .dependsOn(bytes, cli, control, data, dsv, error, io, parse, reflect, ssh, testing, time, trace)
```
Comment (charleso) [inline project/build.scala]: This also need to go under `aggregate` just above

File: mundane-dsv/src/main/scala/com/ambiata/mundane/dsv/DelimitedValuesParser.scala [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+/**
+ * Implementation of a DelimitedValuesParser based on the Apache Commons CSV library
+ */
+case class CommonsDelimitedValuesParser(delimiter: Char, quoteCharacter: Option[Char], escapeCharacter: Option[Char]) extends DelimitedValuesParser {
+  private[this] val format: CSVFormat = {
+    List(
+      (quoteCharacter, (f: CSVFormat, c: Char) => f withQuote c),
+      (escapeCharacter, (f: CSVFormat, c: Char) => f withEscape c)).foldLeft(CSVFormat.newFormat(delimiter)) {
```
Comment (markhibberd) [inline mundane-dsv/src/main/scala/com/ambiata/mundane/dsv/DelimitedValuesParser.scala]: Given this is apache commons we are talking about, I have to ask - is CSVFormat thread safe?

File: mundane-dsv/src/main/scala/com/ambiata/mundane/dsv/DelimitedValuesParser.scala (hunk 1 above)
Comment (markhibberd) [inline mundane-dsv/src/main/scala/com/ambiata/mundane/dsv/DelimitedValuesParser.scala]: I think we will have different implementations, at least until we are comfortable we actually have a super-set of csvish (if one actually exists). I agree that just having createParserXYZ makes a lot of sense, we wouldn't be writing code to support multiple different versions - but different pieces of code - would use different versions, and it is fine (necessary even - from a correctness / complexity management standpoint) for different pieces of code to only work with 1 implementation.

File: mundane-dsv/src/main/scala/com/ambiata/mundane/dsv/DelimitedValuesParser.scala [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+/**
+ * Create a parser for delimited values based on
+ * different implementations and a choice of
+ *
+ *  - delimiter
+ *  - character used for quoting fields
+ *  - character used to escaped control characters (delimiter, quoteCharacter)
+ *
+ */
+object DelimitedValuesParser {
```
Comment (markhibberd) [inline mundane-dsv/src/main/scala/com/ambiata/mundane/dsv/DelimitedValuesParser.scala]: Can we not call this delimited. My preference is for CSV - the standard has moved past the 'C' standing for comma. I really want to differentiate with what we call delimited which is "cleaned" data that should only be split.

File: mundane-dsv/src/main/scala/com/ambiata/mundane/dsv/DelimitedValuesParser.scala [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+  def createParser(delimiter: Char, quoteCharacter: Option[Char], escapeCharacter: Option[Char], implementation: DelimitedValuesParserImplementation): DelimitedValuesParser =
```
Comment (markhibberd) [inline mundane-dsv/src/main/scala/com/ambiata/mundane/dsv/DelimitedValuesParser.scala]: I don't think it is a safe assumption that these things will always be configurable. Probably doesn't matter at the moment, just an observation that I don't think one set of configuration options will make a lot of sense long term.

File: mundane-dsv/src/test/scala/com/ambiata/mundane/dsv/DelimitedValuesParserSpec.scala [hunk 7]
```diff
@@ -0,0 +1,73 @@
+package com.ambiata.mundane.dsv
+
+import com.ambiata.disorder._
+import Arbitraries._
+import org.scalacheck._, Gen._, Arbitrary._
+import org.specs2._
+import org.specs2.matcher.DisjunctionMatchers
+
+class DelimitedValuesParserSpec extends Specification with ScalaCheck with DisjunctionMatchers { def is = s2"""
```
Comment (markhibberd) [inline mundane-dsv/src/test/scala/com/ambiata/mundane/dsv/DelimitedValuesParserSpec.scala]: Needs some benchmarks. Might be worth pulling in the scaffolding off of the benchmark branch @charleso is that possible?

File: mundane-dsv/src/test/scala/com/ambiata/mundane/dsv/DelimitedValuesParserSpec.scala [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+  def genMalformedField: Gen[String] =
+    for {
+      q <- Gen.oneOf("\"") // a non-escaped quote will fail the parse
+      a <- genRegularField
+      b <- genRegularField
+    } yield s"$a$q$b"
+
+
+}
```
Comment (markhibberd) [inline mundane-dsv/src/test/scala/com/ambiata/mundane/dsv/DelimitedValuesParserSpec.scala]: The level of confidence required for this is fairly high, I would probably want to see some larger scale tests, perhaps a code generator and a known corpus of a bunch of documents that we can trivially add to if we come up against unusual edge cases and differences in behaviour between implementations.

File: mundane-dsv/src/main/scala/com/ambiata/mundane/dsv/DelimitedValuesParser.scala [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    implementation match {
+      case CommonsCsv => CommonsDelimitedValuesParser(delimiter, quoteCharacter, escapeCharacter)
+    }
+
+  /** use the CommonsCsv implementation as the default implementation */
+  def delimited(delimiter: Char): DelimitedValuesParser =
```
Comment (markhibberd) [inline mundane-dsv/src/main/scala/com/ambiata/mundane/dsv/DelimitedValuesParser.scala]: This is just thinking out loud, and I know it goes against conventional wisdom, but I would be very tempted to make these not the same type, just make it so they have the same API, I don't know that there is a lot of value in being able to interchange them - and I really _don't_ want to see an end consumer flipping between them - what we have is different use cases demanding different implementations / behaviour - not the desire for a different or configurable behaviour that can cause uncertainty in results.

File: mundane-dsv/src/test/scala/com/ambiata/mundane/dsv/DelimitedValuesParserSpec.scala (hunk 7 above)
Comment (charleso) [inline mundane-dsv/src/test/scala/com/ambiata/mundane/dsv/DelimitedValuesParserSpec.scala]: > Needs some benchmarks. Might be worth pulling in the scaffolding off of the benchmark branch

Yep. Should be trivial. This is the commit you want:

https://github.com/ambiata/mundane/commit/a40fa52f130822c971dd3a34a6d52cfb37cba62d

Comment (charleso) [inline mundane-dsv/src/test/scala/com/ambiata/mundane/dsv/DelimitedValuesParserSpec.scala]: > This is the commit you want:

Let me know if you get stuck Eric. They're easy to write.

**Comment by markhibberd** [comment]
I think the code organisation here works well. Definitely neat, considering what is being pulled in. Very impressed that nothing passes the test - reassuring for all of humanity really. Happy for you to cull implementations that just really don't cut it - but probably worth leaving a few just to do some model-based testing and benchmarking against for any future changes.

**Comment by etorreborre** [comment]
I removed 3 parsers: 

 - CommonsCsv because it was failing too many test cases
 - OpenCsv and AuOpenCsv because they are precursors to SimpleCsv

I kept SimpleCsv also because it can be made thread-safe.

So the remaining parsers are:

 - SimpleCsv (using the `simple-csv` library)
 - BasicCsv (an ugly hand-rolled csv parser)
 - ParboiledCsv (a parser using a PEG grammar with Parboiled)

In terms of performances SimpleCsv is a winner:
```
[info]  0% Scenario{vm=java, trial=0, benchmark=BasicCsvParser} 549278.76 ns; ?=56600.98 ns @ 10 trials
[info] 33% Scenario{vm=java, trial=0, benchmark=ParboiledCsvParser} 650458.40 ns; ?=69101.42 ns @ 10 trials
[info] 67% Scenario{vm=java, trial=0, benchmark=SimpleCsvParser} 102196.87 ns; ?=4228.26 ns @ 10 trials
[info]
[info]          benchmark  us linear runtime
[info]     BasicCsvParser 549 =========================
[info] ParboiledCsvParser 650 ==============================
[info]    SimpleCsvParser 102 ==== 
```

The only drawback with `SimpleCsv` is that it doesn't reject some malformed csv quoted fields which contain unquoted double-quotes like:
``` 
// this is parsed as List(""w\"t\",\"v\"y"")
"\"w\"t\",\"v\"y\"" // toString = "w"t","v"y"
```

**Comment by etorreborre** [comment]
@markhibberd merge ok?

**Comment by markhibberd** [comment]
:japanese_goblin:
